staging/src/k8s.io/cloud-provider/controllers/service/controller.go

```diff
@@ -737,19 +744,29 @@ func (c *Controller) syncNodes(ctx context.Context, workers int) sets.String {
 // load balancers and finished doing it successfully, or didn't try to at all because
 // there's no need. This function returns true if we tried to update load balancers and
 // failed, indicating to the caller that we should try again.
-func (c *Controller) nodeSyncService(svc *v1.Service, oldNodes, newNodes []*v1.Node) bool {
+func (c *Controller) nodeSyncService(svc *v1.Service) bool {
 	const retSuccess = false
 	const retNeedRetry = true
 	if svc == nil || !wantsLoadBalancer(svc) {
 		return retSuccess
 	}
+	newNodes, err := listWithPredicates(c.nodeLister)
+	if err != nil {
+		runtime.HandleError(fmt.Errorf("failed to retrieve node list: %v", err))
+		nodeSyncErrorCount.Inc()
+		return retNeedRetry
+	}
 	newNodes = filterWithPredicates(newNodes, getNodePredicatesForService(svc)...)
-	oldNodes = filterWithPredicates(oldNodes, getNodePredicatesForService(svc)...)
-
+	oldNodes := filterWithPredicates(c.getLastSyncedNodes(svc), getNodePredicatesForService(svc)...)
+	// Store last synced nodes without actually determining if we successfully
+	// synced them or not. Failed node syncs are passed off to retries in the
+	// service queue, so no need to wait. If we don't store it now, we risk
+	// re-syncing all LBs twice, one from another sync in the node sync and
+	// from the service sync
+	c.storeLastSyncedNodes(svc, newNodes)
```

Comment:
these storeLastSyncedNodes and getLastSyncedNodes hold a lock now, so we can't really parallelize this nodeSyncService now

Reply:
otherwise a Service change will call `EnsureLoadBalancer` and a Node change will call `UpdateLoadBalancer` at the same time

Reply:
> That is allowed and shouldn't be a problem as long as the nodes passed are always up-to-date and synchronized between both calls. The issue this PR tries to solve is: synchronizing the nodes which are passed.

sorry for keep asking questions but this controller is always super confusing to me, is that a safe assumption " shouldn't be a problem as long as the nodes passed are always up-to-date and synchronized between both calls"

Reply:
> is that a safe assumption

I think so. We now keep a cache of nodes for each service, which is synchronized and protected by a lock. Given my overall analysis and understanding of this controller: I don't see how the nodes passed to these API calls can now differ, and moreover: how we can end up removing a node which needs to be in the load balancer set. Famous last words, I know.... :)